import all the libraries needed for the analysis


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# import seaborn as sns

In [2]:
# dataset is loaded into the data variable
data = pd.read_csv('D:\datasets\housing.csv')
data.head(100)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0     -122.23     37.88                41.0        880.0           129.0   
1     -122.22     37.86                21.0       7099.0          1106.0   
2     -122.24     37.85                52.0       1467.0           190.0   
3     -122.25     37.85                52.0       1274.0           235.0   
4     -122.25     37.85                52.0       1627.0           280.0   
..        ...       ...                 ...          ...             ...   
95    -122.26     37.80                36.0       5329.0          2477.0   
96    -122.26     37.82                31.0       4596.0          1331.0   
97    -122.26     37.81                29.0        335.0           107.0   
98    -122.26     37.82                22.0       3682.0          1270.0   
99    -122.26     37.82                37.0       3633.0          1085.0   

    population  households  median_income  median_house_value ocean_proximity  
0        322.0       126.0         8.3252            452600.0        NEAR BAY  
1       2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2        496.0       177.0         7.2574            352100.0        NEAR BAY  
3        558.0       219.0         5.6431            341300.0        NEAR BAY  
4        565.0       259.0         3.8462            342200.0        NEAR BAY  
..         ...         ...            ...                 ...             ...  
95      3469.0      2323.0         2.0096            130000.0        NEAR BAY  
96      2048.0      1180.0         2.8345            183800.0        NEAR BAY  
97       202.0        91.0         2.0062            125000.0        NEAR BAY  
98      2024.0      1250.0         1.2185            170000.0        NEAR BAY  
99      1838.0       980.0         2.6104            193100.0        NEAR BAY  

[100 rows x 10 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [4]:
data.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [6]:
# all the rows with null values are removed
data.dropna(inplace=True)

In [36]:
data['ocean_proximity'].value_counts()

<1H OCEAN     9034
INLAND        6496
NEAR OCEAN    2628
NEAR BAY      2270
ISLAND           5
Name: ocean_proximity, dtype: int64

ocean_proximity's label is divided into 5 features in a dataset


In [37]:
# the 5 features of ocean_proximity is concatenate with the dataset b removing ocean_proximity itself
train_data=data.join(pd.get_dummies(data.ocean_proximity)).drop(['ocean_proximity'],axis=1)

In [10]:
train_data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
0           322.0       126.0         8.3252            452600.0          0   
1          2401.0      1138.0         8.3014            358500.0          0   
2           496.0       177.0         7.2574            352100.0          0   
3           558.0       219.0         5.6431            341300.0          0   
4           565.0       259.0         3.8462            342200.0          0   
...           ...         ...            ...                 ...        ...   
20635       845.0       330.0         1.5603             78100.0          0   
20636       356.0       114.0         2.5568             77100.0          0   
20637      1007.0       433.0         1.7000             92300.0          0   
20638       741.0       349.0         1.8672             84700.0          0   
20639      1387.0       530.0         2.3886             89400.0          0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0           0       0         1           0  
1           0       0         1           0  
2           0       0         1           0  
3           0       0         1           0  
4           0       0         1           0  
...       ...     ...       ...         ...  
20635       1       0         0           0  
20636       1       0         0           0  
20637       1       0         0           0  
20638       1       0         0           0  
20639       1       0         0           0  

[20433 rows x 14 columns]

In [12]:
# train_test_split is imported to divide the dataset for training and spliting data

from sklearn.model_selection import train_test_split

X = train_data.drop(['median_house_value'], axis=1)
y = train_data['median_house_value'] 

In [13]:
X


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  <1H OCEAN  INLAND  ISLAND  \
0           322.0       126.0         8.3252          0       0       0   
1          2401.0      1138.0         8.3014          0       0       0   
2           496.0       177.0         7.2574          0       0       0   
3           558.0       219.0         5.6431          0       0       0   
4           565.0       259.0         3.8462          0       0       0   
...           ...         ...            ...        ...     ...     ...   
20635       845.0       330.0         1.5603          0       1       0   
20636       356.0       114.0         2.5568          0       1       0   
20637      1007.0       433.0         1.7000          0       1       0   
20638       741.0       349.0         1.8672          0       1       0   
20639      1387.0       530.0         2.3886          0       1       0   

       NEAR BAY  NEAR OCEAN  
0             1           0  
1             1           0  
2             1           0  
3             1           0  
4             1           0  
...         ...         ...  
20635         0           0  
20636         0           0  
20637         0           0  
20638         0           0  
20639         0           0  

[20433 rows x 13 columns]

In [14]:
y

0        452600.0
1        358500.0
2        352100.0
3        341300.0
4        342200.0
           ...   
20635     78100.0
20636     77100.0
20637     92300.0
20638     84700.0
20639     89400.0
Name: median_house_value, Length: 20433, dtype: float64

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [16]:
X_train

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3339     -122.59     38.78                15.0        764.0           145.0   
14754    -117.05     32.56                17.0        985.0           233.0   
5704     -118.26     34.23                38.0       1107.0           194.0   
14527    -117.14     32.92                 7.0       1308.0           418.0   
1715     -122.31     38.00                29.0       3108.0           534.0   
...          ...       ...                 ...          ...             ...   
879      -122.01     37.56                24.0       2563.0           485.0   
3580     -118.53     34.23                27.0       2131.0           543.0   
10817    -117.95     33.63                17.0       6745.0          1547.0   
2943     -118.97     35.36                31.0       1418.0           306.0   
8099     -118.22     33.81                41.0        726.0           166.0   

       population  households  median_income  <1H OCEAN  INLAND  ISLAND  \
3339        366.0       143.0         3.3750          0       1       0   
14754       811.0       223.0         2.8750          0       0       0   
5704        518.0       195.0         7.5582          1       0       0   
14527       766.0       390.0         3.2151          1       0       0   
1715       1687.0       516.0         4.3333          0       0       0   
...           ...         ...            ...        ...     ...     ...   
879        1174.0       501.0         3.8179          0       0       0   
3580       1065.0       528.0         3.2404          1       0       0   
10817      2688.0      1535.0         3.9917          1       0       0   
2943       1219.0       312.0         1.5743          0       1       0   
8099        602.0       183.0         3.7885          0       0       0   

       NEAR BAY  NEAR OCEAN  
3339          0           0  
14754         0           1  
5704          0           0  
14527         0           0  
1715          1           0  
...         ...         ...  
879           1           0  
3580          0           0  
10817         0           0  
2943          0           0  
8099          0           1  

[16346 rows x 13 columns]

In [17]:
X_test

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20278    -119.20     34.19                19.0       9503.0          1769.0   
3608     -118.46     34.23                16.0       6338.0          1768.0   
12862    -121.39     38.69                30.0       2897.0           506.0   
16603    -120.69     35.62                35.0       3451.0           713.0   
5518     -118.39     33.97                44.0       1097.0           186.0   
...          ...       ...                 ...          ...             ...   
8862     -118.39     34.06                52.0       1213.0           194.0   
885      -121.97     37.55                17.0       4924.0          1247.0   
16694    -120.65     35.12                19.0       2949.0           662.0   
11115    -117.87     33.84                23.0       1678.0           369.0   
16735    -122.40     37.68                41.0       2267.0           486.0   

       population  households  median_income  <1H OCEAN  INLAND  ISLAND  \
20278      6370.0      1718.0         5.0016          0       0       0   
3608       4718.0      1632.0         3.0187          1       0       0   
12862      1508.0       478.0         3.8650          0       1       0   
16603      1550.0       653.0         2.9167          1       0       0   
5518        513.0       185.0         6.2350          1       0       0   
...           ...         ...            ...        ...     ...     ...   
8862        503.0       194.0         8.0095          1       0       0   
885        3080.0      1182.0         3.1680          1       0       0   
16694      1425.0       548.0         2.9615          1       0       0   
11115       912.0       347.0         4.5000          1       0       0   
16735      1045.0       459.0         4.1146          0       0       0   

       NEAR BAY  NEAR OCEAN  
20278         0           1  
3608          0           0  
12862         0           0  
16603         0           0  
5518          0           0  
...         ...         ...  
8862          0           0  
885           0           0  
16694         0           0  
11115         0           0  
16735         1           0  

[4087 rows x 13 columns]

In [18]:
y_train

3339     103100.0
14754    134500.0
5704     263700.0
14527    106300.0
1715     170800.0
           ...   
879      216100.0
3580     230400.0
10817    271600.0
2943      46700.0
8099     156900.0
Name: median_house_value, Length: 16346, dtype: float64

In [19]:
y_test

20278    218500.0
3608     154600.0
12862     88400.0
16603    161700.0
5518     361400.0
           ...   
8862     500001.0
885      189400.0
16694    178100.0
11115    237300.0
16735    272200.0
Name: median_house_value, Length: 4087, dtype: float64

In [20]:
# the model is trained 
from sklearn.linear_model import LinearRegression

reg = LinearRegression()

reg.fit(X_train,y_train)

LinearRegression()

In [22]:
X_test

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20278    -119.20     34.19                19.0       9503.0          1769.0   
3608     -118.46     34.23                16.0       6338.0          1768.0   
12862    -121.39     38.69                30.0       2897.0           506.0   
16603    -120.69     35.62                35.0       3451.0           713.0   
5518     -118.39     33.97                44.0       1097.0           186.0   
...          ...       ...                 ...          ...             ...   
8862     -118.39     34.06                52.0       1213.0           194.0   
885      -121.97     37.55                17.0       4924.0          1247.0   
16694    -120.65     35.12                19.0       2949.0           662.0   
11115    -117.87     33.84                23.0       1678.0           369.0   
16735    -122.40     37.68                41.0       2267.0           486.0   

       population  households  median_income  <1H OCEAN  INLAND  ISLAND  \
20278      6370.0      1718.0         5.0016          0       0       0   
3608       4718.0      1632.0         3.0187          1       0       0   
12862      1508.0       478.0         3.8650          0       1       0   
16603      1550.0       653.0         2.9167          1       0       0   
5518        513.0       185.0         6.2350          1       0       0   
...           ...         ...            ...        ...     ...     ...   
8862        503.0       194.0         8.0095          1       0       0   
885        3080.0      1182.0         3.1680          1       0       0   
16694      1425.0       548.0         2.9615          1       0       0   
11115       912.0       347.0         4.5000          1       0       0   
16735      1045.0       459.0         4.1146          0       0       0   

       NEAR BAY  NEAR OCEAN  
20278         0           1  
3608          0           0  
12862         0           0  
16603         0           0  
5518          0           0  
...         ...         ...  
8862          0           0  
885           0           0  
16694         0           0  
11115         0           0  
16735         1           0  

[4087 rows x 13 columns]

In [21]:
reg.score(X_test, y_test)
print("Linear regression Model", reg.score(X_test, y_test))

Linear regression Model 0.64163616062199


In [39]:
reg.predict([[-122.40,37.68,41.0,2267.0,486.0,1045.0,459.0,4.1146,0,0,0,1,0]])

C:\Users\Lenovo\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([271291.61896051])

In [40]:
y_test.loc[16735]
# y_test

272200.0

our predicted value for given input is 271292 with is approximately close to 272200(y_test)
